pom.xml

```diff
@@ -965,6 +958,79 @@
                         <useIncrementalCompilation>false</useIncrementalCompilation>
                     </configuration>
+                    <executions>
+                        <!-- Refaster rule collections, identified as classes
+                        whose name ends in `Rules`, are compiled to target
+                        Java 8. This causes the OpenRewrite recipe sources
+                        generated from those files by the `rewrite-templating`
+                        annotation processor to also be compiled to target Java
+                        8. The latter is important for compatiblity with the
+                        wider OpenRewrite ecosystem. Note that Refaster rules
+                        can safely reference APIs defined using a higher
+                        bytecode level, as long as they don't use newer
+                        language constructs. -->
+                        <!-- XXX: Once the OpenRewrite baseline is raised to a
+                        more recent Java version, drop and/or deduplicate the
+                        `compilerArgs` configuration. The current configuration
+                        makes sure that the `add-exports` flags are not
+                        specified when targeting Java 8, as they are
+                        unsupported in that context. -->
+                        <!-- XXX: This somehow breaks the Refaster self-check:
+                        `@BeforeTemplate` definitions that contain code
+                        simplified by other rules are no longer flagged. This a
+                        deal-breaker from a maintenance point of view. Figure
+                        out why this happens. -->
+                        <execution>
+                            <id>default-compile</id>
+                            <goals>
+                                <goal>compile</goal>
+                            </goals>
+                            <configuration>
+                                <compilerArgs combine.children="append">
+                                    <arg>--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED</arg>
+                                    <arg>--add-exports=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED</arg>
+                                    <arg>--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED</arg>
+                                    <arg>--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED</arg>
+                                    <arg>--add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED</arg>
+                                    <arg>--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED</arg>
+                                    <arg>--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED</arg>
+                                </compilerArgs>
+                                <excludes>
+                                    <exclude>**/*Rules.java</exclude>
```

Comment:
One might think "why not customize the build only for `*Recipes.java` files?" This appears not to work, because at the start of the compilation those files don't yet exist; this a `maven-compiler-plugin` feature; not a `javac` feature. Open to suggestions, of course!

Reply:
Great to see this start! In my earlier exploration I had wondered if we could use `<proc>only</proc>` to only generate the Java 8 compatible sources first (using Java 17/21), optionally to a different output, and then in a second pass compile those generated sources using Java 8 as target.

Reply:
Interesting idea, to use `<proc>only</proc>`! I just tried some variations of this, but unfortunately that seems to interact badly (different kinds of errors) with subsequent steps (using either `<proc>node</proc>` or `<proc>full</proc>`).

I then tried the "different output" route, where a JAR with a custom classifier (`recipes`) is created: `error-prone-contrib-0.17.1-SNAPSHOT-recipes.jar`. I push that change; would this work for OpenRewrite?

Reply:
A separate jar would work perfectly for us; Might help make for a lighter distribution for EPS itself as well. Thanks a lot!